### PR TITLE
gpio: add driver for MCP23S17 GPIO chip

### PR DESCRIPTION
**Pull request by sknsean**
Re-Open of https://github.com/zephyrproject-rtos/zephyr/pull/11406

Tested on atsamd20 evaluation board connected to a MCP23S17.
I have the I2C version on hand as well MCP23017, wouldn't it be nice to let the exist in the same driver? Maybe called mcp23x17?

Credits to @pveierland for spotting an issue reading inputs correctly.

### CODE REVIEW

File: drivers/gpio/gpio_mcp23s17.c [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+
+	drv_data->spi_cfg.frequency = config->freq;
+	drv_data->spi_cfg.operation = (SPI_OP_MODE_MASTER | SPI_MODE_CPOL |
+				       SPI_MODE_CPHA | SPI_WORD_SET(8) |
+				       SPI_LINES_SINGLE);
+	drv_data->spi_cfg.slave = config->slave;
+
+	k_sem_init(&drv_data->lock, 1, 1);
+
+	dev->driver_api = &api_table;
```
Comment (mnkp) [inline drivers/gpio/gpio_mcp23s17.c]: Not required, done by `DEVICE_AND_API_INIT`.

**Review by mnkp — CHANGES_REQUESTED** [review]
Thanks, looks nice and clean. I do have some blocking comments however.

Once the issues are fixed could you please run the `tests/drivers/gpio/gpio_api_1pin` test. You'll need a board that defines an LED attached to MCP23S17.

And later also `tests/drivers/gpio/gpio_basic_api/`. You'll need to physically short two pins on the board and create a dedicated overlay file. More info in the testcase.

File: drivers/gpio/gpio_mcp23s17.c [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+		}
+
+		drv_data->mcp23s17_cs_ctrl.gpio_pin = config->cs_pin;
+		drv_data->mcp23s17_cs_ctrl.delay = 0;
+
+		drv_data->spi_cfg.cs = &drv_data->mcp23s17_cs_ctrl;
+
+		LOG_DBG("SPI GPIO CS configured on %s:%u",
+			config->cs_dev, config->cs_pin);
+	}
```
[the remark this answers is not in the document]
Comment (sknsean) [reply drivers/gpio/gpio_mcp23s17.c]: Sure?
`config->cs_dev` is not set when DT_INST_0_MICROCHIP_MCP23S17_CS_GPIOS_CONTROLLER is not defined and we do not handle the DT_INST_1_MICROCHIP_MCP23S17_CS_GPIOS_CONTROLLER case.

Or would you like to have it like this:
```
#if defined(DT_INST_0_MICROCHIP_MCP23S17_CS_GPIOS_CONTROLLER) || defined(DT_INST_1_MICROCHIP_MCP23S17_CS_GPIOS_CONTROLLER)
```

File: drivers/gpio/gpio_mcp23s17.c [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+	.reg_cache.ipol = { .all = 0x0 },
+	.reg_cache.gpinten = { .all = 0x0 },
+	.reg_cache.defval = { .all = 0x0 },
+	.reg_cache.intcon = { .all = 0x0 },
+	.reg_cache.iocon = { .all = 0x0 },
+	.reg_cache.gppu = { .all = 0x0 },
+	.reg_cache.intf = { .all = 0x0 },
+	.reg_cache.intcap = { .all = 0x0 },
+	.reg_cache.gpio = { .all = 0x0 },
+	.reg_cache.olat = { .all = 0x0 },
```
[the remark this answers is not in the document]
Comment (sknsean) [reply drivers/gpio/gpio_mcp23s17.c]: Good catch

**Comment by sknsean** [comment]
```
*** Booting Zephyr OS build zephyr-v2.1.0-2185-gbf75a1dd45ac  ***
Running test suite gpio_api_1pin_test
===================================================================
starting test - test_gpio_pin_configure_push_pull
Running test on port=GPIO_E0, pin=1
PASS - test_gpio_pin_configure_push_pull
===================================================================
starting test - test_gpio_pin_configure_single_ended
Running test on port=GPIO_E0, pin=1
When configured as input test pin value is high
Open Drain configuration or Pull Up pin bias is not supported
SKIP - test_gpio_pin_configure_single_ended
===================================================================
starting test - test_gpio_pin_set_get_raw
Running test on port=GPIO_E0, pin=1
PASS - test_gpio_pin_set_get_raw
===================================================================
starting test - test_gpio_pin_set_get
Running test on port=GPIO_E0, pin=1
PASS - test_gpio_pin_set_get
===================================================================
starting test - test_gpio_pin_set_get_active_high
Running test on port=GPIO_E0, pin=1
Step 1: Set logical, get logical and physical pin value
Step 2: Set physical, get logical and physical pin value
PASS - test_gpio_pin_set_get_active_high
===================================================================
starting test - test_gpio_pin_set_get_active_low
Running test on port=GPIO_E0, pin=1
Step 1: Set logical, get logical and physical pin value
Step 2: Set physical, get logical and physical pin value
PASS - test_gpio_pin_set_get_active_low
===================================================================
starting test - test_gpio_pin_toggle
Running test on port=GPIO_E0, pin=1
PASS - test_gpio_pin_toggle
===================================================================
starting test - test_gpio_port_set_masked_get_raw
Running test on port=GPIO_E0, pin=1
PASS - test_gpio_port_set_masked_get_raw
===================================================================
starting test - test_gpio_port_set_masked_get
Running test on port=GPIO_E0, pin=1
PASS - test_gpio_port_set_masked_get
===================================================================
starting test - test_gpio_port_set_masked_get_active_high
Running test on port=GPIO_E0, pin=1
Step 1: Set logical, get logical and physical port value
Step 2: Set physical, get logical and physical port value
PASS - test_gpio_port_set_masked_get_active_high
===================================================================
starting test - test_gpio_port_set_masked_get_active_low
Running test on port=GPIO_E0, pin=1
Step 1: Set logical, get logical and physical port value
Step 2: Set physical, get logical and physical port value
PASS - test_gpio_port_set_masked_get_active_low
===================================================================
starting test - test_gpio_port_set_bits_clear_bits_raw
Running test on port=GPIO_E0, pin=1
PASS - test_gpio_port_set_bits_clear_bits_raw
===================================================================
starting test - test_gpio_port_set_bits_clear_bits
Running test on port=GPIO_E0, pin=1
PASS - test_gpio_port_set_bits_clear_bits
===================================================================
starting test - test_gpio_port_set_clr_bits_raw
Running test on port=GPIO_E0, pin=1
PASS - test_gpio_port_set_clr_bits_raw
===================================================================
starting test - test_gpio_port_set_clr_bits
Running test on port=GPIO_E0, pin=1
PASS - test_gpio_port_set_clr_bits
===================================================================
starting test - test_gpio_port_toggle
Running test on port=GPIO_E0, pin=1
PASS - test_gpio_port_toggle
===================================================================
starting test - test_gpio_int_edge_rising
Running test on port=GPIO_E0, pin=1
Pin interrupt is not supported.
SKIP - test_gpio_int_edge_rising
===================================================================
starting test - test_gpio_int_edge_falling
Running test on port=GPIO_E0, pin=1
Pin interrupt is not supported.
SKIP - test_gpio_int_edge_falling
===================================================================
starting test - test_gpio_int_edge_both
Running test on port=GPIO_E0, pin=1
Pin interrupt is not supported.
SKIP - test_gpio_int_edge_both
===================================================================
starting test - test_gpio_int_edge_to_active
Step 1: Configure pin as active high
Running test on port=GPIO_E0, pin=1
Pin interrupt is not supported.
SKIP - test_gpio_int_edge_to_active
===================================================================
starting test - test_gpio_int_edge_to_inactive
Step 1: Configure pin as active high
Running test on port=GPIO_E0, pin=1
Pin interrupt is not supported.
SKIP - test_gpio_int_edge_to_inactive
===================================================================
starting test - test_gpio_int_level_high
Running test on port=GPIO_E0, pin=1
Pin interrupt is not supported.
SKIP - test_gpio_int_level_high
===================================================================
starting test - test_gpio_int_level_low
Running test on port=GPIO_E0, pin=1
Pin interrupt is not supported.
SKIP - test_gpio_int_level_low
===================================================================
starting test - test_gpio_int_level_active
Step 1: Configure pin as active high
Running test on port=GPIO_E0, pin=1
Pin interrupt is not supported.
SKIP - test_gpio_int_level_active
===================================================================
starting test - test_gpio_int_level_inactive
Step 1: Configure pin as active high
Running test on port=GPIO_E0, pin=1
Pin interrupt is not supported.
SKIP - test_gpio_int_level_inactive
===================================================================
starting test - test_gpio_pin_toggle_visual
Running test on port=GPIO_E0, pin=1
LED ON
LED OFF
LED ON
LED OFF
PASS - test_gpio_pin_toggle_visual
===================================================================
Test suite gpio_api_1pin_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

File: drivers/gpio/gpio_mcp23s17.c [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	if ((flags & GPIO_OPEN_DRAIN) != 0U) {
+		ret = -ENOTSUP;
+		goto done;
+	};
```
Comment (sknsean) [inline drivers/gpio/gpio_mcp23s17.c]: @mnkp good test `gpio_api_1pin` I was missing the devices doesn't support open drain

**Comment by sknsean** [comment]
```
*** Booting Zephyr OS build zephyr-v2.1.0-2185-gae767b4b0839  ***
Running test suite gpio_basic_test
===================================================================
starting test - test_gpio_port
Validate device GPIO_E0
Check GPIO_E0 output 6 connected to input 7
OUT 6 to IN 7 linkage works
- bits_physical
- pin_physical
- check_raw_output_levels
- check_logic_output_levels
- check_input_levels
- bits_logical
- check_pulls
E: MCP23S17: error setting pin pull up/down (-35)
pull-down not supported
PASS - test_gpio_port
===================================================================
starting test - test_gpio_callback_add_remove
test_callback_add_remove not supported
PASS - test_gpio_callback_add_remove
===================================================================
starting test - test_gpio_callback_self_remove
test_callback_self_remove not supported
PASS - test_gpio_callback_self_remove
===================================================================
starting test - test_gpio_callback_enable_disable
test_callback_enable_disable not supported
PASS - test_gpio_callback_enable_disable
===================================================================
starting test - test_gpio_callback_variants
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
interrupts not supported
PASS - test_gpio_callback_variants
===================================================================
Test suite gpio_basic_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```
@mnkp ztest_unit_test(test_gpio_deprecated) is failing... I'm not supporting the deprecated API :)

File: drivers/gpio/gpio_mcp23s17.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mnkp) [reply drivers/gpio/gpio_mcp23s17.c]: Yeah, sorry! I was wrong. This implementation is fine.

**Review by mnkp — APPROVED** [review]
Thanks! Good work!

File: drivers/gpio/gpio_mcp23s17.c [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+		.len = 4,
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+	const struct spi_buf rx_buf[2] = {
+		{
+			.buf = NULL,
+			.len = 2
```
Comment (jhedberg) [inline drivers/gpio/gpio_mcp23s17.c]: I guess this should be `len = 0` since buf is NULL?

File: drivers/gpio/gpio_mcp23s17.c [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+		},
+		{
+			.buf = (u8_t *)&port_data,
+			.len = 2
+		}
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = 2,
```
Comment (jhedberg) [inline drivers/gpio/gpio_mcp23s17.c]: perhaps `ARRAY_SIZE(rx_buf)` instead of the hardcoded `2`?

File: drivers/gpio/gpio_mcp23s17.c [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+			.len = 2,
+		},
+		{
+			.buf = (u8_t *)&port_data,
+			.len = 2,
+		}
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = 2,
```
Comment (jhedberg) [inline drivers/gpio/gpio_mcp23s17.c]: `ARRAY_SIZE(tx_buf)` would be slightly cleaner than a hardcoded `2`

File: drivers/gpio/gpio_mcp23s17.c (hunk 5 above)
Comment (sknsean) [reply drivers/gpio/gpio_mcp23s17.c:59]: Hmm, I could be wrong, but this is the 2 bytes that will be shifted in when the read command is shifted out.
Also seen here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/sensor/lsm6dsl/lsm6dsl_spi.c#L50

Comment (jhedberg) [reply drivers/gpio/gpio_mcp23s17.c:59]: Ok, got it. I (naively) assumed the `len` was only used for an indication for how many bytes are behind the `buf` pointer.

**Comment by galak** [comment]
Rebased on master and updated driver to new DT_INST macros.

**Pull request merged**